### PR TITLE
Assorted FreeBSD fixes

### PR DESCRIPTION
**Pull request by ghost**
## Description

* Allow modules that use evdev to build on FreeBSD.  This will allow building the GBM and Wayland platforms in addition to X11 on FreeBSD
* Some fixes for the Wayland platform on FreeBSD which didn't build anymore with some of the recent changes and waylandpp 0.2.2, because of missing headers for `wayland::presentation_t`, `wayland::xdg_wm_base_t`, and not linking with `libwayland-client-extra++`.
* FreeBSD 12.0 will switch to LLVM's lld on amd64 soon.  Unfortunately `find_soname` will break in that case due to slightly different linker output.  I'm not sure if this is appropriate for Linux so `cmake/scripts/linux/Macros.cmake` was copied for FreeBSD.
* Let Kodi build on ARM architectures on FreeBSD.  Tested on a Raspberry Pi 2 running FreeBSD 12.0-CURRENT a while ago.  Using the UI and a couple of addons worked fine, but when trying to play any audio/video file Kodi simply hangs right now.  But at least this allows one to build Kodi on FreeBSD/arm without jumping through any extra hoops.

I can break any of these out into separate pull requests if necessary.

## Motivation and Context

I maintain the [multimedia/kodi-devel](https://www.freshports.org/multimedia/kodi-devel) port on FreeBSD.  The penultimate goal would be to build it without having to apply custom patches or at least reduce them to a minimum.

## How Has This Been Tested?

As part of `multimedia/kodi-devel` updates on FreeBSD only.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/linux/input/LinuxInputDevices.cpp [hunk 1]
```diff
@@ -403,12 +410,16 @@ int CLinuxInputDevice::KeyboardGetSymbol(unsigned short value)
     break;
 */
   }
-
+#else
```
Comment (Rechi) [inline xbmc/platform/linux/input/LinuxInputDevices.cpp]: return for `TARGET_LINUX`  is now missing

File: cmake/scripts/freebsd/ArchSetup.cmake [hunk 2]
```diff
@@ -1,4 +1,7 @@
 set(ARCH_DEFINES -D_LINUX -DTARGET_POSIX -DTARGET_FREEBSD)
+if(CORE_PLATFORM_NAME_LC STREQUAL rbpi)
+  list(APPEND ARCH_DEFINES -D_ARMEL -DTARGET_RASPBERRY_PI)
```
Comment (Rechi) [inline cmake/scripts/freebsd/ArchSetup.cmake]: would be better fit into `cmake/platform/linux/rbpi.cmake`, like it is already done for aml and gbm

Comment (Rechi) [reply cmake/scripts/freebsd/ArchSetup.cmake]: meant `cmake/platform/freebsd/rbpi.cmake`

**Comment by t6** [comment]
Hmm, the Jenkins build currently fails with this error which doesn't appear to be my fault.  Any chance to fix it?
```
ERROR: $GIT_COMMIT (d4540a9108a5740bd2799448333e3ee87f242ec8) doesn't match currently checked out revision (406db2b2961f355dc3e37112a8ff6de31efa9e82)
```

**Comment by t6** [comment]
I'm going to partially revert the Wayland fixes since they broke the Linux+Wayland build.  It seems there is some sort of problem with `generate-wayland-extra-protocols` on FreeBSD.

File: cmake/platform/freebsd/gbm.cmake [hunk 3]
```diff
@@ -0,0 +1,4 @@
+set(PLATFORM_REQUIRED_DEPS OpenGLES EGL GBM LibDRM)
+set(PLATFORM_OPTIONAL_DEPS VAAPI)
+set(APP_RENDER_SYSTEM gles)
+list(APPEND PLATFORM_DEFINES -DMESA_EGL_NO_X11_HEADERS -DPLATFORM_SETTINGS_FILE=gbm.xml)
```
Comment (Rechi) [inline cmake/platform/freebsd/gbm.cmake]: maybe consider to just include the gbm file from linux like it is done for wayland
https://github.com/xbmc/xbmc/blob/efe30507783ce1fd438df7d0a7d86459dc176585/cmake/platform/freebsd/wayland.cmake

File: cmake/platform/freebsd/rbpi.cmake [hunk 4]
```diff
@@ -0,0 +1,3 @@
+set(PLATFORM_REQUIRED_DEPS OpenGLES EGL MMAL)
```
Comment (Rechi) [inline cmake/platform/freebsd/rbpi.cmake]: same here

**Comment by lrusak** [comment]
For raspberry pi what is the point? AFAIK there I no hw accelerated video decode available on freebsd for the RPi.

**Review by yol — CHANGES_REQUESTED** [review]
There is no need to link `wayland-client-extra++`.

File: cmake/platform/freebsd/rbpi.cmake [hunk 5]
```diff
@@ -0,0 +1,2 @@
+include(cmake/platform/linux/rbpi.cmake)
+list(APPEND ARCH_DEFINES -D_ARMEL -DTARGET_RASPBERRY_PI)
```
Comment (Rechi) [inline cmake/platform/freebsd/rbpi.cmake]: `-D_ARMEL -DTARGET_RASPBERRY_PI` is now in cmake/platform/linux/rbpi.cmake so this can be removed

**Comment by t6** [comment]
> There is no need to link wayland-client-extra++.

Well, Kodi will not link without it:
```
...
/usr/bin/ld: error: undefined symbol: wayland::xdg_toplevel_t::set_fullscreen(wayland::output_t)
>>> referenced by ShellSurfaceXdgShell.cpp
>>>               ShellSurfaceXdgShell.cpp.o:(KODI::WINDOWING::WAYLAND::CShellSurfaceXdgShell::SetFullScreen(wayland::output_t const&, float)) in archive build/windowing/wayland/windowing_WAYLAND.a

/usr/bin/ld: error: undefined symbol: wayland::xdg_toplevel_t::unset_fullscreen()
>>> referenced by ShellSurfaceXdgShell.cpp
>>>               ShellSurfaceXdgShell.cpp.o:(KODI::WINDOWING::WAYLAND::CShellSurfaceXdgShell::SetWindowed()) in archive build/windowing/wayland/windowing_WAYLAND.a
...
```
But this is probably related to the other Wayland trouble I mentioned earlier.  I've removed this from the PR too until I can figure out what the real problem is.  Clearly something else is going wrong. Maybe my `graphics/waylandpp` port is broken in some way.

**Comment by t6** [comment]
Right. This fixes it.

**Review by yol — APPROVED** [review]

**Pull request merged**